cpp/src/arrow/ipc/reader.cc

```diff
@@ -1175,6 +1273,86 @@ Result<std::shared_ptr<RecordBatchFileReader>> RecordBatchFileReader::Open(
   return result;
 }
 
+Future<IpcMessageGenerator::Item> IpcMessageGenerator::operator()() {
+  if (dictionary_index_ < state_->num_dictionaries()) {
+    auto block = FileBlockFromFlatbuffer(
+        state_->footer_->dictionaries()->Get(dictionary_index_++));
+    ARROW_ASSIGN_OR_RAISE(auto fut,
+                          executor_->Submit(&ReadMessageFromBlock, block, state_->file_));
+    return fut;
+  } else if (record_batch_index_ < state_->num_record_batches()) {
+    auto block = FileBlockFromFlatbuffer(
+        state_->footer_->recordBatches()->Get(record_batch_index_++));
+    ARROW_ASSIGN_OR_RAISE(auto fut,
+                          executor_->Submit(&ReadMessageFromBlock, block, state_->file_));
+    return fut;
+  }
+  return Future<Item>::MakeFinished(IterationTraits<Item>::End());
+}
+
+Future<IpcFileRecordBatchGenerator::Item> IpcFileRecordBatchGenerator::operator()() {
+  if (!read_dictionaries_.is_valid()) {
+    std::vector<Future<std::shared_ptr<Message>>> dictionary_messages(
+        state_->num_dictionaries());
+    for (int i = 0; i < state_->num_dictionaries(); i++) {
+      dictionary_messages[i] = message_generator_();
+    }
+    auto dictionaries_read = arrow::All(std::move(dictionary_messages));
```

Comment:
Why `arrow:All` instead of your new fail-fast version?  Seems that would work here.

Reply:
Here, I actually want the results, not just the status. I've renamed this - this future is to just read the dictionary messages, not to actually parse them.